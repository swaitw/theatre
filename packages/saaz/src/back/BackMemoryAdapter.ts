import type {$IntentionalAny, BackStorageAdapter} from '../types'

export class BackMemoryAdapter implements BackStorageAdapter {
  private _state: {
    lists: {
      [key in string]?: Array<{id: string}>
    }
    singularValues: {
      [key in string]?: unknown
    }
  } = {lists: {}, singularValues: {}}

  export(): unknown {
    return this._state
  }

  constructor(state?: unknown) {
    if (state) {
      this._state = state as $IntentionalAny
    }
  }

  async get<T>(key: string): Promise<T | void> {
    return this._state.singularValues[key] as T | void
  }

  async set<T>(key: string, value: T): Promise<void> {
    this._state.singularValues[key] = value
  }

  async pushToList<T extends {id: string}>(
    key: string,
    rows: T[],
  ): Promise<void> {
    if (!this._state.lists[key]) {
      this._state.lists[key] = []
    }
    const list = this._state.lists[key]!
    if (list.some((row) => rows.some((r) => r.id === row.id))) {
      throw new Error(
        `Cannot push to list "${key}" because one or more rows already exist`,
      )
    }
    list.push(...rows)
  }

  async getList<T extends {id: string}>(key: string): Promise<T[]> {
    if (!this._state.lists[key]) {
      return []
    }
    return this._state.lists[key]! as $IntentionalAny
  }

  async pluckFromList<T extends {id: string}>(
    key: string,
    ids: string[],
  ): Promise<Array<T | undefined>> {
    if (!this._state.lists[key]) {
      return []
    }
    const list = this._state.lists[key]!
    return ids.map((id) => {
      const index = list.findIndex((row) => row.id === id)
      if (index === -1) {
        return undefined
      } else {
        const row = list[index]
        list.splice(index, 1)
        return row as T
      }
    })
  }
}
