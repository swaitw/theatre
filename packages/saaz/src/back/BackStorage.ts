import {defer} from '@theatre/utils/defer'
import type {BackStorageAdapter, Transaction} from '../types'

export class BackStorage {
  private _readyD = defer()
  constructor(opts: {dbName: string; storageAdapter: BackStorageAdapter}) {
    this._readyD.resolve(void 0)
  }

  get ready() {
    return this._readyD.promise
  }

  async pushToJournal(opts: {
    dbName: string
    clock: number
    jsonDiff: unknown
    peerId: string
    peerClockFrom: number
    peerClockTo: number
    timestamp: number
    transactions: Transaction[]
  }) {}
}
