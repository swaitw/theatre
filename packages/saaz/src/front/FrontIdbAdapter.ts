import type {IDBPTransaction} from 'idb'
import {openDB} from 'idb'
import type {
  FrontStorageAdapterTransaction,
  FrontStorageAdapter,
} from '../types'

type SingularValue<T> = {
  sessionAndKey: string
  session: string
  value: T
  key: string
}
type ListItem<V> = {
  sessionAndKeyAndId: `${string}/${string}/${string}`
  session: string
  value: V
  key: string
  sessionAndKey: `${string}/${string}`
}

export class FrontIDBAdapter implements FrontStorageAdapter {
  constructor(
    private _idbName: string,
    private _currentSessionName: string,
  ) {}

  private _dbPromise = this._initializeDB()

  async _initializeDB() {
    return openDB(this._idbName, 1, {
      upgrade(db) {
        const singularValueStore = db.createObjectStore('singularValues', {
          keyPath: 'sessionAndKey',
        })

        singularValueStore.createIndex('session', 'session')

        const listsStore = db.createObjectStore('lists', {
          keyPath: 'sessionAndKeyAndId',
        })

        listsStore.createIndex('session', 'session')
        listsStore.createIndex('sessionAndKey', 'sessionAndKey')
      },
    })
  }

  get ready() {
    return this._dbPromise
  }

  async transaction<T>(
    fn: (opts: FrontStorageAdapterTransaction) => Promise<T>,
  ): Promise<T> {
    const db = await this.ready
    const tx = db.transaction(['singularValues', 'lists'], 'readwrite')
    const t = new IDBTransaction(tx)
    return await fn(t)
  }

  async get<T>(key: string, session: string): Promise<T | void> {
    return await this.transaction(({get}) => get(key, session))
  }

  async getList<T extends {id: string}>(
    key: string,
    session: string,
  ): Promise<T[]> {
    return await this.transaction(({getList}) => getList(key, session))
  }
}

class IDBTransaction implements FrontStorageAdapterTransaction {
  constructor(private _tx: IDBPTransaction<unknown, string[], 'readwrite'>) {}

  async get<T>(key: string, session: string): Promise<T | void> {
    const sessionAndKey = `${session}/${key}`
    const store = this._tx.objectStore('singularValues')

    const v = (await store.get(sessionAndKey)) as SingularValue<T>
    return v?.value
  }

  async getAll<T>(key: string): Promise<Record<string, T>> {
    const index = this._tx.objectStore('singularValues').index('session')
    const all: Array<SingularValue<T>> = await index.getAll()

    const record: Record<string, T> = {}

    for (const row of all) {
      record[row.session] = row.value
    }

    return record
  }

  async set<T>(key: string, value: T, session: string): Promise<void> {
    const sessionAndKey = `${session}/${key}`
    const store = this._tx.objectStore('singularValues')
    const s: SingularValue<T> = {
      sessionAndKey: sessionAndKey,
      value,
      session,
      key,
    }

    await store.put(s)
  }

  async deleteSession(session: string): Promise<void> {
    {
      const store = this._tx.objectStore('singularValues')
      const index = store.index('session')
      const sessionIndex = await index.getAllKeys(session)

      for (const key of sessionIndex) {
        await store.delete(key)
      }
    }

    const listsStore = this._tx.objectStore('lists')
    const listsIndex = listsStore.index('session')
    const listsSessionIndex = await listsIndex.getAllKeys(session)

    for (const key of listsSessionIndex) {
      await listsStore.delete(key)
    }
  }

  async pushToList<T extends {id: string}>(
    key: string,
    rows: T[],
    session: string,
  ): Promise<void> {
    const store = this._tx.objectStore('lists')

    for (const row of rows) {
      const listItem: ListItem<T> = {
        key,
        session,
        value: row,
        sessionAndKeyAndId: `${session}/${key}/${row.id}`,
        sessionAndKey: `${session}/${key}`,
      }

      const existingItem = await store.get(listItem.sessionAndKeyAndId)
      if (existingItem) {
        throw new Error(
          `Cannot push to list "${key}" because an entry with id "${row.id}" already exists`,
        )
      }

      // Save the row with key and id properties
      await store.put(listItem)
    }
  }

  async getList<T extends {id: string}>(
    key: string,
    session: string,
  ): Promise<T[]> {
    const store = this._tx.objectStore('lists')
    const keyIndex = store.index('sessionAndKey')
    const sessionAndKey = `${session}/${key}`
    const all = await keyIndex.getAll(sessionAndKey)

    return all.map((s) => s.value)
  }

  async pluckFromList<T extends {id: string}>(
    key: string,
    ids: string[],
    session: string,
  ): Promise<Array<T | undefined>> {
    const store = this._tx.objectStore('lists')

    const results: Array<T | undefined> = []
    for (const id of ids) {
      const sessionAndKeyAndId = `${session}/${key}/${id}`
      const item = await store.get(sessionAndKeyAndId)
      if (item) {
        results.push(item.value as T)
        await store.delete(item)
      } else {
        results.push(undefined)
      }
    }

    return results
  }
}
