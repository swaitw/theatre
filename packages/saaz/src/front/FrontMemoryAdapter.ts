import type {Draft} from 'immer'
import {createDraft, current, finishDraft} from 'immer'
import type {$IntentionalAny, FrontStorageAdapterTransaction} from '../types'
import type {FrontStorageAdapter} from '../types'
import {defer} from '@theatre/utils/defer'

export class FrontMemoryAdapter implements FrontStorageAdapter {
  private _state: {
    sessions: {
      [key in string]?: {
        lists: {
          [key in string]?: Array<{id: string}>
        }
        keyval: {
          [key in string]?: unknown
        }
      }
    }
  } = {sessions: {}}

  private _readyDeferred = defer<void>()
  private _lastTransactionPromise: Promise<void> = Promise.resolve()

  export(): unknown {
    return this._state
  }

  constructor(state?: unknown) {
    if (state) {
      this._state = state as $IntentionalAny
    }
    // nothing to initialize, so just resolve the promise
    this._readyDeferred.resolve()
  }

  get ready() {
    return this._readyDeferred.promise
  }

  async transaction<T>(
    fn: (opts: FrontStorageAdapterTransaction) => Promise<T>,
  ): Promise<T> {
    const deferred = defer<void>()
    const oldTransactionPromise = this._lastTransactionPromise
    this._lastTransactionPromise = deferred.promise
    try {
      await oldTransactionPromise
    } catch (err) {
      // ignore the error. it'll be handled elsewhere
    }

    await this.ready
    const originalState = this._state
    const draft = createDraft(originalState)
    const t = new Transaction(draft)
    try {
      const result = await fn(t)

      this._state = finishDraft(draft)
      return result
    } catch (error) {
      throw error
    } finally {
      deferred.resolve()
    }
  }

  async get<T>(key: string, session: string): Promise<T | void> {
    return await this.transaction(({get}) => get(key, session))
  }

  async getList<T extends {id: string}>(
    key: string,
    session: string,
  ): Promise<T[]> {
    return await this.transaction(({getList}) => getList(key, session))
  }
}

class Transaction implements FrontStorageAdapterTransaction {
  constructor(private _draft: Draft<FrontMemoryAdapter['_state']>) {}

  async get<T>(key: string, session: string): Promise<T | void> {
    return current(this._draft.sessions[session]?.keyval)?.[key] as T | void
  }

  async set<T>(key: string, value: T, session: string): Promise<void> {
    this._draft.sessions[session] ??= {keyval: {}, lists: {}}
    const s = this._draft.sessions[session]!
    s.keyval[key] = value
  }

  async getAll<T>(key: string): Promise<Record<string, T>> {
    const vals: Record<string, T> = {}

    for (const [sessionId, v] of Object.entries(this._draft.sessions)) {
      if (!v) continue
      if (Object.hasOwn(v.keyval, key)) {
        const value: T | undefined = v.keyval[key] as $IntentionalAny
        if (value === undefined) continue
        vals[sessionId] = value
      }
    }

    return vals
  }

  async deleteSession(session: string): Promise<void> {
    delete this._draft.sessions[session]
  }

  async pushToList<T extends {id: string}>(
    key: string,
    rows: T[],
    session: string,
  ): Promise<void> {
    this._draft.sessions[session] ??= {keyval: {}, lists: {}}
    const s = this._draft.sessions[session]!

    if (!s.lists[key]) {
      s.lists[key] = []
    }
    const list = s.lists[key]!
    if (list.some((row) => rows.some((r) => r.id === row.id))) {
      throw new Error(
        `Cannot push to list "${key}" because one or more rows already exist`,
      )
    }
    list.push(...rows)
  }

  async getList<T extends {id: string}>(
    key: string,
    session: string,
  ): Promise<T[]> {
    const s = this._draft.sessions[session]
    if (!s?.lists[key]) {
      return []
    }
    return current(s?.lists)[key]! as $IntentionalAny
  }

  async pluckFromList<T extends {id: string}>(
    key: string,
    ids: string[],
    session: string,
  ): Promise<Array<T | undefined>> {
    this._draft.sessions[session] ??= {keyval: {}, lists: {}}
    const s = this._draft.sessions[session]!

    if (!s.lists[key]) {
      return []
    }
    const list = s.lists[key]!
    return ids.map((id) => {
      const index = list.findIndex((row) => row.id === id)
      if (index === -1) {
        return undefined
      } else {
        const row = list[index]
        list.splice(index, 1)
        return row as T
      }
    })
  }
}
