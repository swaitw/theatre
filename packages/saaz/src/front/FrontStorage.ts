import type {
  $IntentionalAny,
  SessionState,
  FrontStorageAdapterTransaction,
  Transaction,
} from '../types'
import type {FrontStorageAdapter} from '../types'

export class FrontStorage {
  constructor(
    readonly dbName: string,
    private readonly _adapter: FrontStorageAdapter,
  ) {}

  async transaction<T>(fn: (opts: FrontStorageTransaction) => Promise<T>) {
    return await this._adapter.transaction(async (adapterOpts) => {
      const t = new FrontStorageTransactionImpl(adapterOpts)
      return await fn(t)
    })
  }
}

class FrontStorageTransactionImpl {
  constructor(private _adapterTransaction: FrontStorageAdapterTransaction) {}

  async setSessionState(s: SessionState<unknown>, session: string) {
    // TODO: serializing the state every time this changes probably wastes IO.
    // better to create a writeahead log and compact it every once in a while
    await this._adapterTransaction.set('sessionState', s, session)
  }

  async getSessionState(
    session: string,
  ): Promise<SessionState<unknown> | undefined> {
    const v = this._adapterTransaction.get('sessionState', session)
    return v as $IntentionalAny
  }

  async deleteSession(session: string): Promise<void> {
    await this._adapterTransaction.deleteSession(session)
  }

  async getMostRecentlySyncedSessionState(): Promise<
    SessionState<unknown> | undefined
  > {
    const allSessionStates =
      this._adapterTransaction.getAll<SessionState<unknown>>('sessionState')
    let latest: SessionState<unknown> | undefined
    for (const [peerId, sessionState] of Object.entries(allSessionStates)) {
      if (
        typeof sessionState.backendClock === 'number' &&
        sessionState.backendClock > (latest?.backendClock ?? -1)
      ) {
        latest = sessionState
      }
    }

    return latest
  }

  async getOptimisticUpdates(session: string): Promise<Transaction[]> {
    const s = await this._adapterTransaction.getList<{
      id: string
      transaction: Transaction
    }>('optimisticUpdates', session)
    return s.map((t) => t.transaction)
  }

  async pluckOptimisticUpdates(
    transactions: Pick<Transaction, 'peerClock'>[],
    session: string,
  ): Promise<void> {
    await this._adapterTransaction.pluckFromList(
      'optimisticUpdates',
      transactions.map(({peerClock}) => '#' + peerClock),
      session,
    )
  }

  async getAllExistingSessionIds(): Promise<string[]> {
    const all = await this._adapterTransaction.getAll('session')
    return Object.keys(all)
  }

  async pushOptimisticUpdates(
    transactions: Transaction[],
    session: string,
  ): Promise<void> {
    await this._adapterTransaction.pushToList<{
      id: string
      transaction: Transaction
    }>(
      'optimisticUpdates',
      transactions.map((t) => ({
        id: '#' + t.peerClock,
        transaction: t,
      })),
      session,
    )
  }
}

export type FrontStorageTransaction = FrontStorageTransactionImpl
