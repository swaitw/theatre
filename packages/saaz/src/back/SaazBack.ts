import {defer} from '@theatre/utils/defer'
import {applyOptimisticUpdateToState} from '../shared/transactions'
import type {
  BackApplyUpdateOps,
  SaazBackInterface,
  BackStorageAdapter,
  BackGetUpdateSinceClockResult,
  Schema,
  $IntentionalAny,
  PeerSubscribeCallback,
  AllPeersPresenceState,
  PeerPresenceState,
  FullSnapshot,
} from '../types'
import {BackStorage} from './BackStorage'
import type {DebouncedFunc} from 'lodash-es'
import {cloneDeep, throttle} from 'lodash-es'
import {ensureStateIsUptodate as ensureOpStateIsUptodate} from '../shared/utils'
import {Atom} from '@theatre/dataverse'
import deepEqual from '@theatre/utils/deepEqual'

export default class SaazBack implements SaazBackInterface {
  private _dbName: string
  private _storage: BackStorage
  private _readyDeferred = defer<void>()
  private _dbState: FullSnapshot<$IntentionalAny> = {cell: {}, op: {}}
  private _clock: number | null = null
  private _peerStates: {
    [peerId in string]?: {
      lastIncorporatedPeerClock: number
    }
  } = {}
  private _subsribers: Array<PeerSubscribeCallback> = []
  private _schema: Schema<{$schemaVersion: number}>
  private _presenceState: Atom<AllPeersPresenceState> = new Atom({})
  private _schedulePresenseUpdate: DebouncedFunc<() => void>

  constructor(opts: {
    dbName: string
    storageAdapter: BackStorageAdapter
    schema: Schema<$IntentionalAny>
  }) {
    this._schema = opts.schema
    this._dbName = opts.dbName
    this._storage = new BackStorage({
      dbName: opts.dbName,
      storageAdapter: opts.storageAdapter,
    })

    void this._storage.ready.then(() => {
      this._readyDeferred.resolve()
    })

    this._schedulePresenseUpdate = throttle(
      () => {
        this._callSubscribersForPresenceUpdate()
      },
      1000 / 30,
      {leading: true},
    )

    this._schedulePresenseUpdate.cancel()
  }

  get ready() {
    return this._readyDeferred.promise
  }

  get isReady(): boolean {
    return this._readyDeferred.status === 'resolved'
  }

  async getUpdatesSinceClock(opts: {
    clock: number | null
    peerId: string
  }): Promise<BackGetUpdateSinceClockResult> {
    await this._readyDeferred.promise
    return this._getUpdatesSinceClockSync(opts)
  }

  async applyUpdates(
    opts: BackApplyUpdateOps,
  ): Promise<
    ({ok: true} & BackGetUpdateSinceClockResult) | {ok: false; error: unknown}
  > {
    await this._readyDeferred.promise
    return this._applyUpdatesSync(cloneDeep(opts))
  }

  async subscribe(
    opts: {
      peerId: string
    },
    onUpdate: PeerSubscribeCallback,
  ): Promise<() => void> {
    await this._readyDeferred.promise
    return this._subscribeSync(cloneDeep(opts), onUpdate)
  }

  private _getUpdatesSinceClockSync(opts: {
    clock: number | null
    peerId: string
  }): BackGetUpdateSinceClockResult {
    if (!this.ready) {
      throw new Error('Backend is not ready')
    }

    const lastIncorporatedPeerClock =
      this._peerStates[opts.peerId]?.lastIncorporatedPeerClock ?? null

    const clock = this._clock ?? -1
    const {peerId} = opts

    if ((opts.clock ?? -1) === clock) {
      return {hasUpdates: false, peerId, clock, lastIncorporatedPeerClock}
    }

    return {
      hasUpdates: true,
      clock,
      lastIncorporatedPeerClock,
      peerId,
      snapshot: {
        type: 'Snapshot',
        value: this._dbState,
      },
    }
  }

  async getLastIncorporatedPeerClock(opts: {
    peerId: string
  }): Promise<{lastIncorporatedPeerClock: number | null; peerId: string}> {
    await this._readyDeferred.promise
    return {
      peerId: opts.peerId,
      lastIncorporatedPeerClock:
        this._peerStates[opts.peerId]?.lastIncorporatedPeerClock ?? null,
    }
  }

  async closePeer(opts: {peerId: string}): Promise<{ok: boolean}> {
    delete this._peerStates[opts.peerId]
    return {ok: true}
  }

  async updatePresence(opts: {
    peerId: string
    presence: PeerPresenceState
  }): Promise<{ok: true} | {ok: false; error: unknown}> {
    await this._readyDeferred.promise
    return this._updatePresenceSync(cloneDeep(opts))
  }

  private _updatePresenceSync(opts: {
    peerId: string
    presence: PeerPresenceState
  }): {ok: true} | {ok: false; error: unknown} {
    if (!this.ready) {
      throw new Error('Backend is not ready')
    }

    if (!deepEqual(this._presenceState.get()[opts.peerId], opts.presence)) {
      this._presenceState.setByPointer((p) => p[opts.peerId], opts.presence)
      this._schedulePresenseUpdate()
    }

    return {ok: true}
  }

  private _applyUpdatesSync(
    opts: BackApplyUpdateOps,
  ):
    | ({ok: true} & BackGetUpdateSinceClockResult)
    | {ok: false; error: unknown} {
    if (!this.ready) {
      throw new Error('Backend is not ready')
    }

    if (opts.updates.length === 0) {
      return {
        ok: true,
        ...this._getUpdatesSinceClockSync({
          clock: opts.backendClock,
          peerId: opts.peerId,
        }),
      }
    }

    if (!this._peerStates[opts.peerId]) {
      this._peerStates[opts.peerId] = {
        lastIncorporatedPeerClock: -1,
      }
    }
    const peerState = this._peerStates[opts.peerId]!

    // const rebasing = opts.backendClock !== this._clock

    let snapshotSoFar = ensureOpStateIsUptodate(this._dbState, this._schema)
    let lastAcknowledgedClock = peerState.lastIncorporatedPeerClock
    let backendClock = this._clock ?? -1

    // const updatesToIncorporate = []

    for (const update of opts.updates) {
      if (peerState.lastIncorporatedPeerClock >= update.peerClock) {
        continue
      }

      const snapshotBefore = snapshotSoFar
      const [opSnapshotAfter] = applyOptimisticUpdateToState(
        update,
        snapshotBefore,
        this._schema,
        true,
      )
      snapshotSoFar = opSnapshotAfter
      lastAcknowledgedClock = update.peerClock
      backendClock++
    }

    if (lastAcknowledgedClock !== peerState.lastIncorporatedPeerClock) {
      this._dbState = snapshotSoFar
      peerState.lastIncorporatedPeerClock = lastAcknowledgedClock
      this._clock = backendClock
      this._callSubscribersForBackendStateUpdate()
      // TODO: save individual updates to storage
    }

    const s = this._getUpdatesSinceClockSync({
      clock: opts.backendClock,
      peerId: opts.peerId,
    })

    return {
      ok: true,
      ...s,
    }
  }

  private _subscribeSync(
    opts: {peerId: string},
    onUpdate: PeerSubscribeCallback,
  ): () => void {
    this._subsribers.push(onUpdate)
    let unsubscribed = false
    const unsub = () => {
      if (unsubscribed) return
      unsubscribed = true
      const i = this._subsribers.indexOf(onUpdate)
      if (i !== -1) {
        this._subsribers.splice(i, 1)
      }
    }
    return unsub
  }

  private _callSubscribersForBackendStateUpdate() {
    this._schedulePresenseUpdate.cancel()
    for (const sub of this._subsribers) {
      sub({presence: this._presenceState.get(), shouldCheckForUpdates: true})
    }
  }

  private _callSubscribersForPresenceUpdate() {
    this._schedulePresenseUpdate.cancel()
    for (const sub of this._subsribers) {
      sub({presence: this._presenceState.get(), shouldCheckForUpdates: false})
    }
  }
}
