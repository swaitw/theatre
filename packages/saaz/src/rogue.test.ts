import {BackMemoryAdapter} from './back/BackMemoryAdapter'
import SaazBack from './back/SaazBack'
import {FrontMemoryAdapter} from './front/FrontMemoryAdapter'
import {SaazFront} from './front/SaazFront'
import type {Root} from './rogue'
import {change, fromOps, jsonFromCell} from './rogue'
import type { Schema} from './types'

const ahistoricSnapshot: Root = {
  $type: ['map', 'base'],
  $branches: {
    base: {
      $mapProps: {
        foo: {
          $type: ['map', 'base'],
          $branches: {
            base: {
              $boxedValue:
                'some value here, but this will be ignored, because this is an obj register.',
              $mapProps: {
                bar: {
                  $type: ['boxed', 'base'],
                  $branches: {
                    base: {
                      $boxedValue:
                        'some value here. this is an lww register, and it can contain any json value.',
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  },
}

describe(`Rogue`, () => {
  test('setting a non-existing prop', () => {
    const [rep, ops] = change({}, (draft) => {
      expect(draft.a).toBe(undefined)
      draft.a = 1
      expect(draft.a).toBe(1)
    })
    expect(jsonFromCell(rep)).toEqual({a: 1})

    expect([rep, ops]).toMatchSnapshot()
    const [rep2] = fromOps({}, ops)
    expect(rep2).toEqual(rep)
  })
  test('overriding an existing prop with the same value', () => {
    const [rep1] = change({}, (draft) => {
      draft.a = 1
    })
    const [rep2, ops] = change(rep1, (draft) => {
      expect(draft.a).toBe(1)
      draft.a = 1
      expect(draft.a).toBe(1)
    })
    expect(rep1).toBe(rep2)
    expect(ops).toEqual([])
  })
  test('overriding an existing prop', () => {
    const [rep1] = change({}, (draft) => {
      draft.a = 1
    })
    const [rep2, ops] = change(rep1, (draft) => {
      expect(draft.a).toBe(1)
      draft.a = 2
      expect(draft.a).toBe(2)
    })
    expect(jsonFromCell(rep2)).toEqual({a: 2})
    expect(ops).toHaveLength(1)
    expect(ops).toMatchSnapshot()
    const [rep3] = fromOps(rep1, ops)
    expect(rep3).toEqual(rep2)
  })
  test('setting a non-existing prop to an object', () => {
    const [rep, ops] = change({}, (draft) => {
      expect(draft.a).toBe(undefined)
      draft.a = {b: 1}
      expect(draft.a).toEqual(draft.a)
      expect(draft.a).toEqual({b: 1})
    })
    expect(jsonFromCell(rep)).toEqual({a: {b: 1}})

    expect([rep, ops]).toMatchSnapshot()
    const [rep2] = fromOps({}, ops)
    expect(rep2).toEqual(rep)
  })

  test('setting an existing prop to an object', () => {
    const [rep] = change({}, (draft) => {
      draft.a = {b: 1}
    })
    expect(jsonFromCell(rep)).toEqual({a: {b: 1}})
    const [rep2, ops2] = change(rep, (draft) => {
      expect(draft.a).toEqual({b: 1})
      draft.a = {b: 2}
      expect(draft.a).toEqual({b: 2})
    })

    expect(jsonFromCell(rep2)).toEqual({a: {b: 2}})
    const [rep3] = fromOps(rep, ops2)
    expect(rep3).toEqual(rep2)
  })
  test('setting an existing prop from an object', () => {
    const [rep] = change({}, (draft) => {
      draft.a = {b: 1}
    })
    expect(jsonFromCell(rep)).toEqual({a: {b: 1}})
    const [rep2, ops2] = change(rep, (draft) => {
      expect(draft.a).toEqual({b: 1})
      draft.a = {c: 1}
      expect(draft.a).toEqual({c: 1})
    })

    expect(jsonFromCell(rep2)).toEqual({a: {c: 1}})
    const [rep3] = fromOps(rep, ops2)
    expect(rep3).toEqual(rep2)
  })

  test(`merging defaults`, () => {
    const [, ops1_1] = change({}, (draft) => {
      draft.a = {aStep: 1, foo: 'setBy1', obj: {objA: 'true'}}
    })
    const [, ops2_1] = change({}, (draft) => {
      draft.a = {bStep: 1, foo: 'setBy2', obj: {objB: 'true'}}
    })

    const merge1 = fromOps({}, [...ops2_1, ...ops1_1])[0]

    const _11 = jsonFromCell(merge1)
    expect(_11).toMatchSnapshot()
    // console.log(_11)
  })

  function scenario(
    name: string,
    steps: Record<
      string,
      (draft: any, lastSnapshot: any, lastOps: any[]) => void
    >,
  ) {
    describe(name, () => {
      type StepResult = {
        json: any
        ops: any[]
        backwardOps: any[]
        rep: any
        next?: StepResult
        prev?: StepResult
      }
      let last: StepResult = {
        json: {},
        backwardOps: [],
        ops: [],
        rep: {},
      }
      const byStep: Record<string, StepResult> = {}

      for (const [stepName, fn] of Object.entries(steps)) {
        test(stepName, () => {
          const prev: StepResult = {...last}
          const [rep, ops, backwardOps] = change(prev.rep, (draft) => {
            fn(draft, prev.json, prev.ops)
          })
          const stepResult: StepResult = {
            rep,
            ops,
            backwardOps,
            json: jsonFromCell(rep),
            prev,
          }
          last.next = stepResult
          last = stepResult
          byStep[stepName] = stepResult
        })
      }

      // i = 0 so that we skip the first step, which is the initial state
      for (let i = 1; i < Object.keys(steps).length; i++) {
        const prevStepName = Object.keys(steps)[i - 1]
        const stepName = Object.keys(steps)[i]
        test(`${prevStepName} => ${stepName}`, () => {
          const stepResult = byStep[stepName]
          const [rep] = fromOps(stepResult.prev!.rep, stepResult.ops)
          expect(rep).toEqual(stepResult.rep)
        })

        test(`${prevStepName} <= ${stepName}`, () => {
          const stepResult = byStep[stepName]
          const [rep] = fromOps(stepResult.rep, stepResult.backwardOps)
          const s = jsonFromCell(rep)
          // note that as opposed to the previous test, we're not comparing cells, we're
          // comparing snapshots. This is because the cells are not guaranteed to be the
          // same when undoing a change, but the snapshots are.
          expect(s).toEqual(stepResult.prev!.json)
        })
      }
    })
  }

  scenario('scenario 1', {
    step1: (draft) => {
      expect(draft.a).toBe(undefined)
      draft.a = 1
      expect(draft.a).toBe(1)
    },
    step2: (_, snapshot, ops) => {
      expect(snapshot).toEqual({a: 1})
    },
  })
  scenario('scenario 2', {
    step1: (draft) => {
      draft.a = {a1: {a11: 1}}
      expect(draft.a.a1).toEqual({a11: 1})
      draft.a.a1.a11 = 2
      expect(draft.a).toEqual({a1: {a11: 2}})
    },
    step2: (_, snapshot, ops) => {
      expect(snapshot).toEqual({a: {a1: {a11: 2}}})
    },
  })
  scenario('scenario 3', {
    step1: (draft) => {
      draft.a = {a1: {a11: 1}}
      expect(draft.a.a1).toEqual({a11: 1})
      draft.a = {b: 1}
      expect(draft.a).toEqual({b: 1})
    },
    step2: (draft, snapshot, ops) => {
      expect(draft.a).toEqual({b: 1})
      draft.a = 1
    },
    step3: (draft, snapshot) => {
      expect(snapshot).toEqual({a: 1})
    },
  })
  describe(`saaz integration`, () => {
    test(`test`, async () => {
      type State = {
        $schemaVersion: number
        count: number
      }

      const schema: Schema<State> = {
        version: 1,
        // migrateOp(state: $IntentionalAny) {},
        // migrateCell(s) {},
        generators: {},
        editors: {
          increaseBy(state: State, generators: {}, opts: {by: number}) {
            state.count += opts.by
          },
        },
        opShape: null as any as State,
        cellShape: null as any as {},
      }
      const backend = new SaazBack({
        schema,
        dbName: 'test',
        storageAdapter: new BackMemoryAdapter(),
      })
      const saaz = new SaazFront({
        schema,
        dbName: 'test',
        peerId: '1',
        storageAdapter: new FrontMemoryAdapter(),
        backend,
      })

      saaz.tx((editors) => {})

      saaz.teardown()
    })
  })
})
