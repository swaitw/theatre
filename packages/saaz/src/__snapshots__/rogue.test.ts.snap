// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rogue merging defaults 1`] = `
{
  "a": {
    "aStep": 1,
    "bStep": 1,
    "foo": "setBy1",
    "obj": {
      "objA": "true",
      "objB": "true",
    },
  },
}
`;

exports[`Rogue overriding an existing prop 1`] = `
[
  {
    "branchName": "base",
    "path": [
      [
        "base",
        "a",
      ],
    ],
    "type": "SetBoxedValue",
    "value": 2,
  },
]
`;

exports[`Rogue setting a non-existing prop 1`] = `
[
  {
    "$branches": {
      "base": {
        "$mapProps": {
          "a": {
            "$branches": {
              "base": {
                "$boxedValue": 1,
              },
            },
            "$type": [
              "boxed",
              "base",
            ],
          },
        },
      },
    },
    "$type": [
      "map",
      "base",
    ],
  },
  [
    {
      "path": [
        [
          "base",
          "a",
        ],
      ],
      "type": "ChangeType",
      "value": [
        "boxed",
        "base",
      ],
    },
    {
      "branchName": "base",
      "path": [
        [
          "base",
          "a",
        ],
      ],
      "type": "SetBoxedValue",
      "value": 1,
    },
  ],
]
`;

exports[`Rogue setting a non-existing prop to an object 1`] = `
[
  {
    "$branches": {
      "base": {
        "$mapProps": {
          "a": {
            "$branches": {
              "base": {
                "$mapProps": {
                  "b": {
                    "$branches": {
                      "base": {
                        "$boxedValue": 1,
                      },
                    },
                    "$type": [
                      "boxed",
                      "base",
                    ],
                  },
                },
              },
            },
            "$type": [
              "map",
              "base",
            ],
          },
        },
      },
    },
    "$type": [
      "map",
      "base",
    ],
  },
  [
    {
      "path": [
        [
          "base",
          "a",
        ],
      ],
      "type": "ChangeType",
      "value": [
        "map",
        "base",
      ],
    },
    {
      "path": [
        [
          "base",
          "a",
        ],
        [
          "base",
          "b",
        ],
      ],
      "type": "ChangeType",
      "value": [
        "boxed",
        "base",
      ],
    },
    {
      "branchName": "base",
      "path": [
        [
          "base",
          "a",
        ],
        [
          "base",
          "b",
        ],
      ],
      "type": "SetBoxedValue",
      "value": 1,
    },
  ],
]
`;
